gateway/middleware.go

```diff
@@ -311,17 +312,43 @@ func (t BaseMiddleware) ApplyPolicies(session *user.SessionState) error {
 	}
 
 	didQuota, didRateLimit, didACL, didComplexity := make(map[string]bool), make(map[string]bool), make(map[string]bool), make(map[string]bool)
-	policies := session.PolicyIDs()
-
-	for _, polID := range policies {
-		t.Gw.policiesMu.RLock()
-		policy, ok := t.Gw.policiesByID[polID]
-		t.Gw.policiesMu.RUnlock()
-		if !ok {
-			err := fmt.Errorf("policy not found: %q", polID)
-			t.Logger().Error(err)
-			return err
+	policyIDs := session.PolicyIDs()
```

Comment:
Could L315-331 land in a `func (s *SessionData) PolicyIDs() ([]string, error)` please. This enables protections for concurrent map access over the session.MetaData map, e.g. can implement locking/mutex. This behaviour should be local to *SessionData.

Similarly, the second block should be in a function akin to `func (gw *Gateway) getPolicies([]string) ([]user.Policy, error)`.

Inlined code is hard to test, separate functions are better.